consensus/peer_state.go

```diff
@@ -0,0 +1,522 @@
+package consensus
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	cstypes "github.com/tendermint/tendermint/consensus/types"
+	"github.com/tendermint/tendermint/libs/bits"
+	tmjson "github.com/tendermint/tendermint/libs/json"
+	"github.com/tendermint/tendermint/libs/log"
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+	"github.com/tendermint/tendermint/p2p"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+	tmtime "github.com/tendermint/tendermint/types/time"
+)
+
+var (
+	ErrPeerStateHeightRegression = errors.New("error peer state height regression")
+	ErrPeerStateInvalidStartTime = errors.New("error peer state invalid startTime")
+)
+
+// peerStateStats holds internal statistics for a peer.
+type peerStateStats struct {
+	Votes      int `json:"votes"`
+	BlockParts int `json:"block_parts"`
+}
+
+func (pss peerStateStats) String() string {
+	return fmt.Sprintf("peerStateStats{votes: %d, blockParts: %d}", pss.Votes, pss.BlockParts)
+}
+
+// PeerState contains the known state of a peer, including its connection and
+// threadsafe access to its PeerRoundState.
+// NOTE: THIS GETS DUMPED WITH rpc/core/consensus.go.
+// Be mindful of what you Expose.
+type PeerState struct {
+	peerID p2p.NodeID
+	logger log.Logger
+
+	// NOTE: Modify below using setters, never directly.
+	mtx     tmsync.RWMutex
+	running bool
+	PRS     cstypes.PeerRoundState `json:"round_state"`
+	Stats   *peerStateStats        `json:"stats"`
+
+	broadcastWG sync.WaitGroup
+	closer      *tmsync.Closer
+}
+
+// NewPeerState returns a new PeerState for the given node ID.
+func NewPeerState(logger log.Logger, peerID p2p.NodeID) *PeerState {
+	return &PeerState{
+		peerID: peerID,
+		logger: logger,
+		closer: tmsync.NewCloser(),
+		PRS: cstypes.PeerRoundState{
+			Round:              -1,
+			ProposalPOLRound:   -1,
+			LastCommitRound:    -1,
+			CatchupCommitRound: -1,
+		},
+		Stats: &peerStateStats{},
+	}
+}
+
+// SetRunning sets the running state of the peer.
+func (ps *PeerState) SetRunning(v bool) {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	ps.running = v
+}
+
+// IsRunning returns true if a PeerState is considered running where multiple
+// broadcasting goroutines exist for the peer.
+func (ps *PeerState) IsRunning() bool {
+	ps.mtx.RLock()
+	defer ps.mtx.RUnlock()
+
+	return ps.running
```

Comment:
I'm a bit worried about having two different signals for shutdown here. Can we just rely on `ps.closer` for `IsRunning()` too?

Reply:
How would that work? Use `ps.closer.Done()`? Wouldn't that block?

I don't think the current method is too bad.

Reply:
You'd do a non-blocking read to check whether it was closed or not.

Either way, since this isn't concurrent it's not that important -- but if it was, there's be issues due to non-atomicity so you'd have to be super-careful about the ordering of operations and checks.